File: matlab/make_tables_from_h5.m
% MAKE_TABLES_FROM_H5  Build light tables (MS/SAR) pointing into LCZ42 HDF5 files.
% root = 'data/lcz42'; with training.h5, validation.h5, testing.h5
%
% Produces:
%   tables_MS.mat  -> train_MS, val_MS, test_MS
%   tables_SAR.mat -> train_SAR, val_SAR, test_SAR
%
% Each table has variables: Path (string), Label (categorical 1..17),
% Index (1-based), Modality ("MS"/"SAR")
%
% Matteo Rambaldi — Thesis utilities

function make_tables_from_h5(root)

    arguments
        root (1,1) string
    end
    
    tr = fullfile(root,'training.h5');
    te = fullfile(root,'testing.h5');
    assert(isfile(tr) && isfile(te), 'Missing H5 files in %s', root);
    
    % --- read counts and labels (labels are one-hot N x 17) ---
    [Ntr, lab_tr] = getN_and_labels(tr);
    [Nte, lab_te] = getN_and_labels(te);
    
    % --- build Sentinel-2 (MS) tables ---
    train_MS = table(repmat(string(tr),Ntr,1), lab_tr, (1:Ntr)', repmat("MS",Ntr,1), ...
        'VariableNames',{'Path','Label','Index','Modality'});
    test_MS  = table(repmat(string(te),Nte,1), lab_te, (1:Nte)', repmat("MS",Nte,1), ...
        'VariableNames',{'Path','Label','Index','Modality'});
    
    % --- build Sentinel-1 (SAR) tables (same rows; Modality changes) ---
    train_SAR = train_MS; train_SAR.Modality(:) = "SAR";
    test_SAR  = test_MS;  test_SAR.Modality(:)  = "SAR";
    
    % --- save ---
    save(fullfile(root,'tables_MS.mat'),'train_MS','test_MS','-v7.3');
    save(fullfile(root,'tables_SAR.mat'),'train_SAR','test_SAR','-v7.3');
    
    fprintf('Saved:\n  %s\n  %s\n', fullfile(root,'tables_MS.mat'), fullfile(root,'tables_SAR.mat'));
end

% ======================================================================
%                               HELPERS
% ======================================================================

% Shapes of the files:
%   /sen1 : [8, 32, 32, N]
%   /sen2 : [10,32, 32, N]
%   /label: [17, N]
function [N, labels_cat] = getN_and_labels(h5path)
    d1 = h5info(h5path,'/sen1').Dataspace.Size;
    d2 = h5info(h5path,'/sen2').Dataspace.Size;
    dl = h5info(h5path,'/label').Dataspace.Size;
    
    assert(numel(d1)==4 && d1(1)==8  && d1(2)==32 && d1(3)==32, '/sen1 dims unexpected: %s', mat2str(d1));
    assert(numel(d2)==4 && d2(1)==10 && d2(2)==32 && d2(3)==32, '/sen2 dims unexpected: %s', mat2str(d2));
    assert(numel(dl)==2 && dl(1)==17, '/label dims unexpected: %s', mat2str(dl));
    
    N = d1(4);
    assert(d2(4)==N && dl(2)==N, 'N mismatch in %s (sen1/sen2/label).', h5path);
    
    L = h5read(h5path,'/label');   % [17, N]
    labels_cat = onehot_to_cat(L);
end

function C = onehot_to_cat(L)
    % L: [17, N]
    if size(L,1)==17
        [~,idx] = max(L, [], 1);  % 1×N
        idx = idx.';              % N×1
    else
        error('Unexpected label size: %s', mat2str(size(L)));
    end
    C = categorical(idx);
end

File: resnet18_ensembles/scripts/make_tables_from_h5.py
import h5py
import numpy as np
import pandas as pd
from scipy.io import savemat
import os

def make_tables_from_h5(data_dir):
    """
    Build tables_MS.mat and tables_SAR.mat from HDF5 dataset.
    Compatible with MATLAB make_tables_from_h5.m
    """
    train_h5 = os.path.join(data_dir, "training.h5")
    test_h5  = os.path.join(data_dir, "testing.h5")

    def build_table(h5_path, modality):
        with h5py.File(h5_path, "r") as f:
            n_samples = f["label"].shape[0]
            print(f"[INFO] {modality} → Found {n_samples} samples in {h5_path}")

        path_list = [h5_path] * n_samples
        label = np.argmax(h5py.File(h5_path, "r")["label"][:], axis=1) + 1  # 1-based
        index = np.arange(1, n_samples + 1)
        modality_list = [modality] * n_samples

        df = pd.DataFrame({
            "Path": path_list,
            "Label": label,
            "Index": index,
            "Modality": modality_list,
        })
        return df

    # --- Build tables ---
    print("[INFO] Building tables for MS (/sen2)...")
    train_MS = build_table(train_h5, "MS")
    test_MS  = build_table(test_h5,  "MS")

    print("[INFO] Building tables for SAR (/sen1)...")
    train_SAR = build_table(train_h5, "SAR")
    test_SAR  = build_table(test_h5,  "SAR")

    # --- Save as MATLAB .mat ---
    out_ms = os.path.join(data_dir, "tables_MS.mat")
    out_sar = os.path.join(data_dir, "tables_SAR.mat")

    savemat(out_ms, {
    "train_MS": train_MS.values,
    "test_MS": test_MS.values
    })
    savemat(out_sar, {
        "train_SAR": train_SAR.values,
        "test_SAR": test_SAR.values
    })

    print(f"Saved {out_ms} and {out_sar} (MATLAB-compatible)")

if __name__ == "__main__":
    make_tables_from_h5("../../data/lcz42")